File: src/main/java/com/smarthome/backend/server/api/services/modules/DenonModuleService.java
package com.smarthome.backend.server.api.services.modules;

import java.io.IOException;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.gson.Gson;
import com.smarthome.backend.server.api.ApiRouter;
import com.smarthome.backend.server.api.modules.heos.denon.DenonModule;
import com.smarthome.backend.server.db.DatabaseManager;

/**
 * Service für Denon-Module-API-Endpunkte.
 */
public class DenonModuleService {
    private static final Logger logger = LoggerFactory.getLogger(DenonModuleService.class);
    private static final Gson gson = new Gson();
    
    private final DenonModule denonModule;
    
    public DenonModuleService(DatabaseManager databaseManager) {
        this.denonModule = new DenonModule(databaseManager);
    }
    
    public void handleRequest(com.sun.net.httpserver.HttpExchange exchange, String method, String path) throws IOException {
        logger.debug("DenonModuleService: {} {}", method, path);
        
        if (path.equals("/modules/denon/devices/discover")) {
            if ("GET".equals(method)) {
                discoverDevices(exchange);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/denon/devices/[^/]+/setVolume")) {
            if ("POST".equals(method)) {
                setVolume(exchange, path);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/denon/devices/[^/]+/setOn")) {
            if ("POST".equals(method)) {
                setOn(exchange, path);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/denon/devices/[^/]+/setOff")) {
            if ("POST".equals(method)) {
                setOff(exchange, path);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/denon/devices/[^/]+/setPlayState")) {
            if ("POST".equals(method)) {
                setPlayState(exchange, path);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/denon/devices/[^/]+/setMute")) {
            if ("POST".equals(method)) {
                setMute(exchange, path);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/denon/devices/[^/]+/playNext")) {
            if ("POST".equals(method)) {
                playNext(exchange, path);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else if (path.matches("/modules/denon/devices/[^/]+/playPrevious")) {
            if ("POST".equals(method)) {
                playPrevious(exchange, path);
            } else {
                ApiRouter.sendResponse(exchange, 405, gson.toJson(Map.of("error", "Method not allowed")));
            }
        } else {
            ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Endpoint not found")));
        }
    }
    
    private void discoverDevices(com.sun.net.httpserver.HttpExchange exchange) throws IOException {
        denonModule.discoverDevices(exchange);
    }
    
    private void setVolume(com.sun.net.httpserver.HttpExchange exchange, String path) throws IOException {
        try {
            String deviceId = ApiRouter.extractPathParam(path, "/modules/denon/devices/{deviceId}/setVolume");
            if (deviceId == null || deviceId.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültige Device ID")));
                return;
            }
            
            String requestBody = ApiRouter.readRequestBody(exchange);
            if (requestBody == null || requestBody.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültiger Body")));
                return;
            }
            
            @SuppressWarnings("unchecked")
            Map<String, Object> request = gson.fromJson(requestBody, Map.class);
            if (request == null || !request.containsKey("volume")) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Volume-Parameter fehlt")));
                return;
            }
            
            Object volumeObj = request.get("volume");
            int volume;
            if (volumeObj instanceof Number) {
                volume = ((Number) volumeObj).intValue();
            } else if (volumeObj instanceof String) {
                volume = Integer.parseInt((String) volumeObj);
            } else {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültiger Volume-Wert")));
                return;
            }
            
            boolean success = denonModule.setVolume(deviceId, volume);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Lautstärke", e);
            ApiRouter.sendResponse(exchange, 500, gson.toJson(Map.of("error", "Fehler beim Setzen der Lautstärke: " + e.getMessage())));
        }
    }
    
    private void setOn(com.sun.net.httpserver.HttpExchange exchange, String path) throws IOException {
        try {
            String deviceId = ApiRouter.extractPathParam(path, "/modules/denon/devices/{deviceId}/setOn");
            if (deviceId == null || deviceId.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültige Device ID")));
                return;
            }
            
            boolean success = denonModule.setPlayState(deviceId, "play");
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Einschalten des Geräts", e);
            ApiRouter.sendResponse(exchange, 500, gson.toJson(Map.of("error", "Fehler beim Einschalten des Geräts: " + e.getMessage())));
        }
    }
    
    private void setOff(com.sun.net.httpserver.HttpExchange exchange, String path) throws IOException {
        try {
            String deviceId = ApiRouter.extractPathParam(path, "/modules/denon/devices/{deviceId}/setOff");
            if (deviceId == null || deviceId.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültige Device ID")));
                return;
            }
            
            boolean success = denonModule.setPlayState(deviceId, "stop");
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Ausschalten des Geräts", e);
            ApiRouter.sendResponse(exchange, 500, gson.toJson(Map.of("error", "Fehler beim Ausschalten des Geräts: " + e.getMessage())));
        }
    }
    
    private void setPlayState(com.sun.net.httpserver.HttpExchange exchange, String path) throws IOException {
        try {
            String deviceId = ApiRouter.extractPathParam(path, "/modules/denon/devices/{deviceId}/setPlayState");
            if (deviceId == null || deviceId.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültige Device ID")));
                return;
            }
            
            String requestBody = ApiRouter.readRequestBody(exchange);
            if (requestBody == null || requestBody.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültiger Body")));
                return;
            }
            
            @SuppressWarnings("unchecked")
            Map<String, Object> request = gson.fromJson(requestBody, Map.class);
            if (request == null || !request.containsKey("state")) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "State-Parameter fehlt")));
                return;
            }
            
            String state = (String) request.get("state");
            if (state == null || (!state.equals("play") && !state.equals("pause") && !state.equals("stop"))) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültiger State-Wert (muss 'play', 'pause' oder 'stop' sein)")));
                return;
            }
            
            boolean success = denonModule.setPlayState(deviceId, state);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Setzen des Wiedergabestatus", e);
            ApiRouter.sendResponse(exchange, 500, gson.toJson(Map.of("error", "Fehler beim Setzen des Wiedergabestatus: " + e.getMessage())));
        }
    }
    
    private void setMute(com.sun.net.httpserver.HttpExchange exchange, String path) throws IOException {
        try {
            String deviceId = ApiRouter.extractPathParam(path, "/modules/denon/devices/{deviceId}/setMute");
            if (deviceId == null || deviceId.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültige Device ID")));
                return;
            }
            
            String requestBody = ApiRouter.readRequestBody(exchange);
            if (requestBody == null || requestBody.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültiger Body")));
                return;
            }
            
            @SuppressWarnings("unchecked")
            Map<String, Object> request = gson.fromJson(requestBody, Map.class);
            if (request == null || !request.containsKey("mute")) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Mute-Parameter fehlt")));
                return;
            }
            
            Object muteObj = request.get("mute");
            boolean mute;
            if (muteObj instanceof Boolean) {
                mute = (Boolean) muteObj;
            } else if (muteObj instanceof String) {
                mute = Boolean.parseBoolean((String) muteObj);
            } else {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültiger Mute-Wert")));
                return;
            }
            
            boolean success = denonModule.setMute(deviceId, mute);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Stummschaltung", e);
            ApiRouter.sendResponse(exchange, 500, gson.toJson(Map.of("error", "Fehler beim Setzen der Stummschaltung: " + e.getMessage())));
        }
    }
    
    private void playNext(com.sun.net.httpserver.HttpExchange exchange, String path) throws IOException {
        try {
            String deviceId = ApiRouter.extractPathParam(path, "/modules/denon/devices/{deviceId}/playNext");
            if (deviceId == null || deviceId.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültige Device ID")));
                return;
            }
            
            boolean success = denonModule.playNext(deviceId);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Abspielen des nächsten Titels", e);
            ApiRouter.sendResponse(exchange, 500, gson.toJson(Map.of("error", "Fehler beim Abspielen des nächsten Titels: " + e.getMessage())));
        }
    }
    
    private void playPrevious(com.sun.net.httpserver.HttpExchange exchange, String path) throws IOException {
        try {
            String deviceId = ApiRouter.extractPathParam(path, "/modules/denon/devices/{deviceId}/playPrevious");
            if (deviceId == null || deviceId.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400, gson.toJson(Map.of("error", "Ungültige Device ID")));
                return;
            }
            
            boolean success = denonModule.playPrevious(deviceId);
            if (success) {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(Map.of("success", true)));
            } else {
                ApiRouter.sendResponse(exchange, 404, gson.toJson(Map.of("error", "Gerät nicht gefunden")));
            }
        } catch (Exception e) {
            logger.error("Fehler beim Abspielen des vorherigen Titels", e);
            ApiRouter.sendResponse(exchange, 500, gson.toJson(Map.of("error", "Fehler beim Abspielen des vorherigen Titels: " + e.getMessage())));
        }
    }
    
}

File: src/main/node/com/smarthome/backend/server/api/services/modules/denonModule.service.ts
import { Router } from "express";
import { DenonModuleManager } from "../../modules/heos/denon/denonModuleManager.js";
import { logger } from "../../../../logger.js";
import type { ServerDeps } from "../../server.js";

export function createDenonModuleRouter(deps: ServerDeps) {
  const router = Router();
  const denonModule = new DenonModuleManager(deps.databaseManager, deps.deviceManager, deps.eventManager);
  deps.deviceManager.registerModuleManager(denonModule);

  router.get("/devices/discover", async (_req, res) => {
    try {
      const devices = await denonModule.discoverDevices();
      res.status(200).json(devices);
    } catch (error) {
      logger.error({ error }, "Fehler beim Discover von Denon-Geräten");
      res.status(500).json({ error: "Fehler beim Discover von Denon-Geräten" });
    }
  });

  router.post("/devices/:deviceId/setVolume", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const rawVolume = (req.body ?? {}).volume;
      let volume: number | null = null;
      if (typeof rawVolume === "number") {
        volume = rawVolume;
      } else if (typeof rawVolume === "string" && rawVolume.trim().length > 0) {
        const parsed = Number(rawVolume);
        if (!Number.isNaN(parsed)) volume = parsed;
      }
      if (volume == null) {
        res.status(400).json({ error: "Volume-Parameter fehlt" });
        return;
      }
      const success = await denonModule.setVolume(deviceId, volume);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der Lautstärke");
      res.status(500).json({ error: "Fehler beim Setzen der Lautstärke" });
    }
  });

  router.post("/devices/:deviceId/setOn", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const success = await denonModule.setPlayState(deviceId, "play");
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Einschalten des Geräts");
      res.status(500).json({ error: "Fehler beim Einschalten des Geräts" });
    }
  });

  router.post("/devices/:deviceId/setOff", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const success = await denonModule.setPlayState(deviceId, "stop");
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Ausschalten des Geräts");
      res.status(500).json({ error: "Fehler beim Ausschalten des Geräts" });
    }
  });

  router.post("/devices/:deviceId/setPlayState", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const state = String((req.body ?? {}).state ?? "");
      if (!["play", "pause", "stop"].includes(state)) {
        res.status(400).json({ error: "Ungültiger State-Wert (muss 'play', 'pause' oder 'stop' sein)" });
        return;
      }
      const success = await denonModule.setPlayState(deviceId, state);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen des Wiedergabestatus");
      res.status(500).json({ error: "Fehler beim Setzen des Wiedergabestatus" });
    }
  });

  router.post("/devices/:deviceId/setMute", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const rawMute = (req.body ?? {}).mute;
      let mute: boolean | null = null;
      if (typeof rawMute === "boolean") {
        mute = rawMute;
      } else if (typeof rawMute === "string") {
        if (rawMute === "true" || rawMute === "false") {
          mute = rawMute === "true";
        }
      }
      if (mute == null) {
        res.status(400).json({ error: "Mute-Parameter fehlt" });
        return;
      }
      const success = await denonModule.setMute(deviceId, mute);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der Stummschaltung");
      res.status(500).json({ error: "Fehler beim Setzen der Stummschaltung" });
    }
  });

  router.post("/devices/:deviceId/playNext", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const success = await denonModule.playNext(deviceId);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Abspielen des naechsten Titels");
      res.status(500).json({ error: "Fehler beim Abspielen des naechsten Titels" });
    }
  });

  router.post("/devices/:deviceId/playPrevious", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const success = await denonModule.playPrevious(deviceId);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Abspielen des vorherigen Titels");
      res.status(500).json({ error: "Fehler beim Abspielen des vorherigen Titels" });
    }
  });

  router.post("/devices/:deviceId/setVolumeStart", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const raw = (req.body ?? {}).volumeStart;
      const volumeStart = typeof raw === "number" ? raw : (typeof raw === "string" ? Number(raw) : NaN);
      if (Number.isNaN(volumeStart) || volumeStart < 0 || volumeStart > 50) {
        res.status(400).json({ error: "Parameter volumeStart fehlt oder ungültig (0–50)" });
        return;
      }
      const success = await denonModule.setVolumeStart(deviceId, volumeStart);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen von Volume-Start");
      res.status(500).json({ error: "Fehler beim Setzen von Volume-Start" });
    }
  });

  router.post("/devices/:deviceId/setVolumeMax", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const raw = (req.body ?? {}).volumeMax;
      const volumeMax = typeof raw === "number" ? raw : (typeof raw === "string" ? Number(raw) : NaN);
      if (Number.isNaN(volumeMax) || volumeMax < 40 || volumeMax > 98) {
        res.status(400).json({ error: "Parameter volumeMax fehlt oder ungültig (40–98)" });
        return;
      }
      const success = await denonModule.setVolumeMax(deviceId, volumeMax);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen von Volume-Max");
      res.status(500).json({ error: "Fehler beim Setzen von Volume-Max" });
    }
  });

  router.post("/devices/:deviceId/setSource", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const sourceIndex = String((req.body ?? {}).sourceIndex ?? "").trim();
      const rawSelected = (req.body ?? {}).selected;
      const selected = rawSelected === true || rawSelected === "true";
      if (!sourceIndex) {
        res.status(400).json({ error: "Parameter sourceIndex fehlt" });
        return;
      }
      const success = await denonModule.setSource(deviceId, sourceIndex, selected);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der aktiven Quelle");
      res.status(500).json({ error: "Fehler beim Setzen der aktiven Quelle" });
    }
  });

  router.post("/devices/:deviceId/setZonePower", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const zoneName = String((req.body ?? {}).zoneName ?? "").trim();
      const rawPower = (req.body ?? {}).power;
      const power = rawPower === true || rawPower === "true";
      if (!zoneName) {
        res.status(400).json({ error: "Parameter zoneName fehlt" });
        return;
      }
      const success = await denonModule.setZonePower(deviceId, zoneName, power);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der Zonen-Power");
      res.status(500).json({ error: "Fehler beim Setzen der Zonen-Power" });
    }
  });

  router.post("/devices/:deviceId/setSubwooferPower", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const subwooferId = String((req.body ?? {}).subwooferId ?? "").trim();
      const rawPower = (req.body ?? {}).power;
      let power: boolean | null = null;
      if (typeof rawPower === "boolean") {
        power = rawPower;
      } else if (typeof rawPower === "string") {
        if (rawPower === "true" || rawPower === "false") {
          power = rawPower === "true";
        }
      }
      if (!subwooferId || power == null) {
        res.status(400).json({ error: "Parameter subwooferId oder power fehlt bzw. ungültig" });
        return;
      }
      const success = await denonModule.setSubwooferPower(deviceId, subwooferId, power);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen der Subwoofer-Power");
      res.status(500).json({ error: "Fehler beim Setzen der Subwoofer-Power" });
    }
  });

  router.post("/devices/:deviceId/setSubwooferLevel", async (req, res) => {
    try {
      const deviceId = req.params.deviceId;
      if (!deviceId) {
        res.status(400).json({ error: "Ungültige Device ID" });
        return;
      }
      const subwooferId = String((req.body ?? {}).subwooferId ?? "").trim();
      const rawLevel = (req.body ?? {}).level;
      const level = typeof rawLevel === "number" ? rawLevel : (typeof rawLevel === "string" ? Number(rawLevel) : NaN);
      if (!subwooferId || Number.isNaN(level)) {
        res.status(400).json({ error: "Parameter subwooferId oder level fehlt bzw. ungültig" });
        return;
      }
      const success = await denonModule.setSubwooferLevel(deviceId, subwooferId, level);
      res.status(success ? 200 : 404).json(success ? { success: true } : { error: "Gerät nicht gefunden" });
    } catch (error) {
      logger.error({ error }, "Fehler beim Setzen des Subwoofer-Pegels");
      res.status(500).json({ error: "Fehler beim Setzen des Subwoofer-Pegels" });
    }
  });

  return router;
}

